From 6d74e25040c64c31bfb41b3789b7751685148efc Mon Sep 17 00:00:00 2001
From: "lei.chen" <lei.chen@amlogic.com>
Date: Tue, 15 Aug 2023 11:15:35 +0800
Subject: [PATCH] BT: amlogic fix net/bluetooth driver [1/1]

PD#SWPL-128975
PD#SWPL-118242
PD#SWPL-265980

Problem:
1.rdk kernel 5.15 bt not support deep sleep wake-up
2.rdk kernel 5.15 need user confirm

Solution:
1.set CONFIG_AMLOGIC_LINUX_BT as linux in yocto
2.use same ble create connection parameters as android
3.filter hci bt fw log
4.add CONFIG_AMLOGIC_LINUX_BT_SMP
5.add CONFIG_AMLOGIC_LINUX_BT_SUPPORT_WAKEUP

Verify:
T5W T3X

Change-Id: Id900189d0a1b66b4c42cf0ca2eb6a9e479b9440c
Signed-off-by: lei.chen <lei.chen@amlogic.com>
Signed-off-by: Zhu Lv <zhu.lv@amlogic.com>
---
 net/bluetooth/hci_core.c |  7 +++++++
 net/bluetooth/hci_sync.c | 18 ++++++++++++++++++
 net/bluetooth/smp.c      |  8 ++++++++
 3 files changed, 33 insertions(+)

diff --git a/net/bluetooth/hci_core.c b/net/bluetooth/hci_core.c
index 3369dc105000..e137cf5a4bb1 100644
--- a/net/bluetooth/hci_core.c
+++ b/net/bluetooth/hci_core.c
@@ -4062,7 +4062,14 @@ static void hci_rx_work(struct work_struct *work)
 		switch (hci_skb_pkt_type(skb)) {
 		case HCI_EVENT_PKT:
 			BT_DBG("%s Event packet", hdev->name);
+#if IS_ENABLED(CONFIG_AMLOGIC_LINUX_BT)
+			if (skb->data[0] != 0x62)	/* filter hci bt fw log */
+				hci_event_packet(hdev, skb);
+			else
+				kfree_skb(skb);
+#else
 			hci_event_packet(hdev, skb);
+#endif
 			break;
 
 		case HCI_ACLDATA_PKT:
diff --git a/net/bluetooth/hci_sync.c b/net/bluetooth/hci_sync.c
index f79b38603205..9ab4a042097e 100644
--- a/net/bluetooth/hci_sync.c
+++ b/net/bluetooth/hci_sync.c
@@ -6151,7 +6151,9 @@ static int hci_pause_scan_sync(struct hci_dev *hdev)
  */
 int hci_suspend_sync(struct hci_dev *hdev)
 {
+#ifndef CONFIG_AMLOGIC_LINUX_BT_SUPPORT_WAKEUP
 	int err;
+#endif
 
 	/* If marked as suspended there nothing to do */
 	if (hdev->suspended)
@@ -6173,6 +6175,7 @@ int hci_suspend_sync(struct hci_dev *hdev)
 	hci_pause_scan_sync(hdev);
 
 	if (hci_conn_count(hdev)) {
+#ifndef CONFIG_AMLOGIC_LINUX_BT_SUPPORT_WAKEUP
 		/* Soft disconnect everything (power off) */
 		err = hci_disconnect_all_sync(hdev, HCI_ERROR_REMOTE_POWER_OFF);
 		if (err) {
@@ -6181,6 +6184,9 @@ int hci_suspend_sync(struct hci_dev *hdev)
 			hci_resume_sync(hdev);
 			return err;
 		}
+#else
+		bt_dev_info(hdev, "amlogic skip disconnect while suspend");
+#endif
 
 		/* Update event mask so only the allowed event can wakeup the
 		 * host.
@@ -6436,8 +6442,14 @@ static void set_ext_conn_params(struct hci_conn *conn,
 	p->scan_window = cpu_to_le16(hdev->le_scan_window_connect);
 	p->conn_interval_min = cpu_to_le16(conn->le_conn_min_interval);
 	p->conn_interval_max = cpu_to_le16(conn->le_conn_max_interval);
+#if IS_ENABLED(CONFIG_AMLOGIC_LINUX_BT)
+	// Specify the same create connection parameters as for Android
+	p->conn_latency = 0;
+	p->supervision_timeout = 500;	// 5S
+#else
 	p->conn_latency = cpu_to_le16(conn->le_conn_latency);
 	p->supervision_timeout = cpu_to_le16(conn->le_supv_timeout);
+#endif
 	p->min_ce_len = cpu_to_le16(0x0000);
 	p->max_ce_len = cpu_to_le16(0x0000);
 }
@@ -6579,8 +6591,14 @@ static int hci_le_create_conn_sync(struct hci_dev *hdev, void *data)
 	cp.own_address_type = own_addr_type;
 	cp.conn_interval_min = cpu_to_le16(conn->le_conn_min_interval);
 	cp.conn_interval_max = cpu_to_le16(conn->le_conn_max_interval);
+#if IS_ENABLED(CONFIG_AMLOGIC_LINUX_BT)
+	// Specify the same create connection parameters as for Android
+	cp.conn_latency = 0;
+	cp.supervision_timeout = 500;  // 5S
+#else
 	cp.conn_latency = cpu_to_le16(conn->le_conn_latency);
 	cp.supervision_timeout = cpu_to_le16(conn->le_supv_timeout);
+#endif
 	cp.min_ce_len = cpu_to_le16(0x0000);
 	cp.max_ce_len = cpu_to_le16(0x0000);
 
diff --git a/net/bluetooth/smp.c b/net/bluetooth/smp.c
index a31971fe2fd7..b565d8313d01 100644
--- a/net/bluetooth/smp.c
+++ b/net/bluetooth/smp.c
@@ -884,6 +884,13 @@ static int tk_request(struct l2cap_conn *conn, u8 remote_oob, u8 auth,
 	    hcon->io_capability == HCI_IO_NO_INPUT_OUTPUT)
 		smp->method = JUST_WORKS;
 
+#if IS_ENABLED(CONFIG_AMLOGIC_LINUX_BT_SMP)
+	/* If Just Works, Continue with Zero TK */
+	if (smp->method == JUST_WORKS) {
+		set_bit(SMP_FLAG_TK_VALID, &smp->flags);
+		return 0;
+	}
+#else
 	/* If Just Works, Continue with Zero TK and ask user-space for
 	 * confirmation */
 	if (smp->method == JUST_WORKS) {
@@ -896,6 +903,7 @@ static int tk_request(struct l2cap_conn *conn, u8 remote_oob, u8 auth,
 		set_bit(SMP_FLAG_WAIT_USER, &smp->flags);
 		return 0;
 	}
+#endif
 
 	/* If this function is used for SC -> legacy fallback we
 	 * can only recover the just-works case.
-- 
2.42.0

